### PR TITLE
reference/performance: add doc for join reorder

### PR DESCRIPTION
**Pull request by winoros**
### What is the related PR or file link(s)? 

- [ ] Reference link(s):
- [ ] This PR is to align with:
- [ ] N/A (not applicable)

### Which TiDB version(s) does your changes apply to? 

- [ ] All active versions: dev, v3.0, v2.1, v3.1
- [x] dev (the latest development version)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] All active and inactive versions
- [ ] N/A (not applicable)

**Note:** If your changes apply to multiple TiDB versions, make sure you update the documents in the **corresponding** version folders such as "dev", "v3.0", "v2.1" and "v3.1" in this PR.

- [ ] Updated one version first. Will update other versions after I get two LGTMs.

### CODE REVIEW

**Comment by TomShawn** [comment]
@winoros Please fix the CI failure, thanks.

**Comment by yikeke** [comment]
Images should be put into the `/media` folder. @winoros

File: reference/performance/sql-optimizer-overview.md [hunk 1]
```diff
@@ -18,6 +18,7 @@ category: reference
 - 谓词下推
 - 分区裁剪
 - TopN 和 Limit 下推
+- [Join Reorder](./join-reorder.md)
```
Comment (yikeke) [inline reference/performance/sql-optimizer-overview.md]:
```suggestion
- [Join Reorder](/reference/performance/join-reorder.md)
```
Links are given in a unified way in our documentation - the whole path should be written. 

Please update all the links in this PR, then the lint check will pass. @winoros

File: reference/performance/join-reorder.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
+---
+title: TiDB 当前的 Join Reorder 算法简介
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
title: Join Reorder 算法简介
```

File: reference/performance/join-reorder.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+category: reference
+---
+
+# Join Reorder
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
# Join Reorder 算法简介
```
一级标题需要与上面元信息里的 title 保持一致。

File: reference/performance/join-reorder.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+在实际的的业务场景中，多个表的 Join 语句是非常常见的。而 Join 的执行效果会和各个表参与 Join 的顺序有关系。如 `select * from t, t1, t2 where t.a=t1.a and t2.a=t1.a`，这个 SQL 中可能的执行顺序有`t 和 t1 先做 Join，然后再和 t2 做 Join`以及`t1 和 t2 先做 Join，然后再和 t 做 Join` 两种情况。根据 `t` 和 `t2` 的数据量及数据分布，这两种执行顺序会有不同的性能表现。
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
在实际的业务场景中，多个表的 Join 语句是很常见的，而 Join 的执行效果和各个表参与 Join 的顺序有关系。如 `select * from t, t1, t2 where t.a=t1.a and t2.a=t1.a`，这个 SQL 中可能的执行顺序有“t 和 t1 先做 Join，然后再和 t2 做 Join”以及“t1 和 t2 先做 Join，然后再和 t 做 Join”两种情况。根据 `t` 和 `t2` 的数据量及数据分布，这两种执行顺序会有不同的性能表现。
```

File: reference/performance/join-reorder.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+因此需要优化器实现一种决定 Join 顺序的算法。目前 TiDB 中的算法是贪心算法。
+
+# 贪心算法简介
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
## Join Reorder 算法实例
```
一篇文档只能有一个一级标题，即全文的标题。

File: reference/performance/join-reorder.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+以三个表 `t1, t2, t3` 的 Join 为例。首先我们会从拿到所有参与 Join 的节点，将他们按照行数多少从少到多排序
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
以三个表 t1、t2、t3 的 Join 为例。首先获取所有参与 Join 的节点，将所有节点按照行数多少，从少到多进行排序。
```

File: reference/performance/join-reorder.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+![pic1](/media/join-reorder-1.png)
+
+之后我们选定其中最小的表。和其他两个表一次做 Join，看输出的结果集大小，选择更小的一侧作为选择。
+
+![pic2](/media/join-reorder-2.png)
+
+然后进入下一轮的选择，如果这时是四个表那么就是继续比较输出结果集的大小进行选择。这里只有三个表，因此就直接得到了最终的 Join 结果。
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
然后进入下一轮的选择，如果这时是四个表，那么就继续比较输出结果集的大小，进行选择。这里只有三个表，因此就直接得到了最终的 Join 结果。
```

File: reference/performance/join-reorder.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+因此需要优化器实现一种决定 Join 顺序的算法。目前 TiDB 中的算法是贪心算法。
+
+# 贪心算法简介
+
+以三个表 `t1, t2, t3` 的 Join 为例。首先我们会从拿到所有参与 Join 的节点，将他们按照行数多少从少到多排序
+
+![pic1](/media/join-reorder-1.png)
+
+之后我们选定其中最小的表。和其他两个表一次做 Join，看输出的结果集大小，选择更小的一侧作为选择。
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
之后选定其中最小的表，将其与其他两个表分别做一次 Join，观察输出的结果集大小，选择其中结果更小的一对。
```

File: reference/performance/join-reorder.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+
+![pic2](/media/join-reorder-2.png)
+
+然后进入下一轮的选择，如果这时是四个表那么就是继续比较输出结果集的大小进行选择。这里只有三个表，因此就直接得到了最终的 Join 结果。
+
+![pic3](/media/join-reorder-3.png)
+
+这个就是当前 TiDB 中使用的 Join reorder 算法。
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
以上就是当前 TiDB 中使用的 Join reorder 算法。
```

File: reference/performance/join-reorder.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
+---
+title: TiDB 当前的 Join Reorder 算法简介
+category: reference
+---
+
+# Join Reorder
+
+在实际的的业务场景中，多个表的 Join 语句是非常常见的。而 Join 的执行效果会和各个表参与 Join 的顺序有关系。如 `select * from t, t1, t2 where t.a=t1.a and t2.a=t1.a`，这个 SQL 中可能的执行顺序有`t 和 t1 先做 Join，然后再和 t2 做 Join`以及`t1 和 t2 先做 Join，然后再和 t 做 Join` 两种情况。根据 `t` 和 `t2` 的数据量及数据分布，这两种执行顺序会有不同的性能表现。
+
+因此需要优化器实现一种决定 Join 顺序的算法。目前 TiDB 中的算法是贪心算法。
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
因此优化器需要实现一种决定 Join 顺序的算法。目前 TiDB 中使用的算法是 Join Reorder 算法，又称贪心算法。
```
Join Reorder 算法 和 贪心算法 是一个意思吗？文章的内容需要统一一下这两个词的用法。

File: reference/performance/join-reorder.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+![pic2](/media/join-reorder-2.png)
+
+然后进入下一轮的选择，如果这时是四个表那么就是继续比较输出结果集的大小进行选择。这里只有三个表，因此就直接得到了最终的 Join 结果。
+
+![pic3](/media/join-reorder-3.png)
+
+这个就是当前 TiDB 中使用的 Join reorder 算法。
+
+# 限制
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
## Join reorder 算法限制
```

File: reference/performance/join-reorder.md [hunk 12]
```diff
@@ -0,0 +1,35 @@
+---
```
Comment (yikeke) [inline reference/performance/join-reorder.md]: Please add this doc link into `TOC.md`.

File: reference/performance/join-reorder.md [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+在实际的的业务场景中，多个表的 Join 语句是非常常见的。而 Join 的执行效果会和各个表参与 Join 的顺序有关系。如 `select * from t, t1, t2 where t.a=t1.a and t2.a=t1.a`，这个 SQL 中可能的执行顺序有`t 和 t1 先做 Join，然后再和 t2 做 Join`以及`t1 和 t2 先做 Join，然后再和 t 做 Join` 两种情况。根据 `t` 和 `t2` 的数据量及数据分布，这两种执行顺序会有不同的性能表现。
+
+因此需要优化器实现一种决定 Join 顺序的算法。目前 TiDB 中的算法是贪心算法。
+
+# 贪心算法简介
+
+以三个表 `t1, t2, t3` 的 Join 为例。首先我们会从拿到所有参与 Join 的节点，将他们按照行数多少从少到多排序
+
+![pic1](/media/join-reorder-1.png)
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
![join-reorder-1](/media/join-reorder-1.png)
```
For SEO reasons, it's better to make the image link description self-explanatory. Please update all image link descriptions in this document.

File: reference/performance/join-reorder.md (hunk 6 above)
Comment (francis0407) [inline reference/performance/join-reorder.md]: 这里用了 `t1, t2, t3` ，要不要把上面的例子也改成 `t1, t2, t3`?

File: reference/performance/join-reorder.md (hunk 4 above)
Comment (francis0407) [inline reference/performance/join-reorder.md]: > 而 Join 的执行效果和各个表参与 Join 的顺序

“效果”用“效率”或者“性能”更好？（“效果”有一种结果会不一样的感觉）

**Comment by yikeke** [comment]
ping @winoros  Please resolve comments.

File: reference/performance/join-reorder.md (hunk 13 above)
Comment (yikeke) [reply reference/performance/join-reorder.md]: 文档中还有几个其他的 link description 需要更新。 @winoros

File: reference/performance/join-reorder.md [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+## Join Reorder 算法实例
+
+以三个表 t1、t2、t3 的 Join 为例。首先获取所有参与 Join 的节点，将所有节点按照行数多少，从少到多进行排序。
+
+![join-reorder-1](/media/join-reorder-1.png)
+
+之后选定其中最小的表，将其与其他两个表分别做一次 Join，观察输出的结果集大小，选择其中结果更小的一对。
+
+![pic2](/media/join-reorder-2.png)
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
![join-reorder-2](/media/join-reorder-2.png)
```

File: reference/performance/join-reorder.md [hunk 15]
```diff
@@ -0,0 +1,35 @@
...
+
+然后进入下一轮的选择，如果这时是四个表，那么就继续比较输出结果集的大小，进行选择。这里只有三个表，因此就直接得到了最终的 Join 结果。
+
+![pic3](/media/join-reorder-3.png)
```
Comment (yikeke) [inline reference/performance/join-reorder.md]:
```suggestion
![join-reorder-3](/media/join-reorder-3.png)
```

**Comment by ran-huang** [comment]
@yikeke Shall we add a link to join-reorder.md in TOC?

**Comment by yikeke** [comment]
> @yikeke Shall we add a link to join-reorder.md in TOC?

现在在 TOC 里还没有添加 4.0 文档的链接，等到时官网新增 4.0 入口了我统一加到 TOC 里。

**Pull request merged**